lib/os/queued_operation.c

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/queued_operation.h>
+#include <stdint.h>
+
+/* States used in the manager state field. */
+enum state {
+	/* Service is not active.
+	 *
+	 * Transitions to STARTING on queued_operation_submit().
+	 */
+	ST_OFF = 0,
+
+	/* Service is being started.
+	 *
+	 * This is a transient state while an associated on-off
+	 * service request is incomplete.  Transitions to IDLE and
+	 * reschedules on successful start, and ERROR on failure to
+	 * start.
+	 */
+	ST_STARTING,
+
+	/* Service is active with no known operations.
+	 *
+	 * This is a transient substate of an implicit ON state.  The
+	 * machine will transition to NOTIFYING or STOPPING within
+	 * the current mutex region
+	 */
+	ST_IDLE,
+
+	/* The manager is invoking process() to notify the service of
+	 * a new operation or a transition to idle.
+	 *
+	 * Transitions to PROCESSING if an operation was passed and
+	 * queued_operation_finalize() has not been invoked before
+	 * process() returns to the manager.
+	 *
+	 * Transitions to IDLE if an operation was not passed and the
+	 * manager queue remains empty when process() returns to the
+	 * manager.
+	 *
+	 * Re-runs select in all other cases.
+	 */
+	ST_NOTIFYING,
+
+	/* A new operation has been identified and the service
+	 * process() function will be/is/has-been invoked on it.
+	 *
+	 * Transitions to FINALIZING when queued_operation_finalize()
+	 * is invoked.
+	 */
+	ST_PROCESSING,
+
+	/* An operation that was processing is being finalized.
+	 *
+	 * Re-selects after finalization and any containing notifying
+	 * completes.
+	 */
+	ST_FINALIZING,
+
+	/* Service is being started.
+	 *
+	 * This is a transient state while an associated on-off
+	 * service request is pending.
+	 */
+	ST_STOPPING,
+
+	/* Service is in an error state. */
+	ST_ERROR,
```

Comment:
seems that there is no way to get out of error state. How to reinitialize the manager?